python/social_bot/teacher_tasks.py

```diff
@@ -226,6 +226,7 @@ def __init__(self,
                  random_range=10.0,
                  random_goal=False,
                  sparse_reward=True,
+                 step_time=0.1,
```

Comment:
actually step_time is not used in this GoalWithDistractionTask task. but other tasks may need playground to tell them step_time. Any better way for this?

Reply:
Nevermind, we can keep it this way. interface of tasks should be more unified.
Because sim step time and substeps might be different for each agent and env, and can be override in gin config, in some task this option is used to get the time span of 1 step from the env (e.g., reward shaping, caculate normalized speed to goal from distance to goal of 2 steps will need this).  So we need something like Task(stepsize=**) in env. But not all env need this.

Reply:
Thanks for explaining Jiangtao! It seems strange to have a configurable named step_time used in computing speed.  Shouldn't speed be fully determined by the physics?  Maybe consider changing the name of this variable, and maybe also the formula to make it more natural?

Reply:
step_time is the time passed by an env step. 
`step_time = substeps * substep_time`
substep_time is the simulator's accuracy, i.e., max_step_size in world file. Typically step_time is 0.1 (10Hz roll-out freq), and max_step_size is 0.001. 

While speed is numerical normalized, since there are 100 simulator sub-steps depends on the configuration, speed to goal (the agent speed projected on the direction to goal) calculated by simulator is from the last 0.001s sub-step, does not reflect the real progress made to goal in the entire 0.1s env step, especially when the agent is in an accelerating progress. (distance_to_goal-pre_distance_to_goal)/step_time is more accurate here.

```
normalized_progress = (distance_to_goal - pre_distance_to_goal) / initial_distance
speed_to_goal =  (distance_to_goal - pre_distance_to_goal) / step_time
```
I'll add some comments for this.

Reply:
I see, step_time and substeps are configured, and substep_time is inferred from the two?

If step_time is the elapsed time per each env step, then shouldn't it be part of the env, instead of the task?

Reply:
yes.  substeps is inferred from the step_time and substep_time. it is part of the env, but the task need this. So when create the task, this parameter should be passed.

Reply:
Since max_step_size is specified in the world file, and NUM_SIMULATION_STEPS (substeps) is typically specified in specific tasks, it is indeed better to compute the step_time value based on these two variables. This calculation can be done outside of this task as you mentioned.
For now, perhaps set the default value to an invalid value and add assertion on value when using it?

Reply:
There is a maximal max_step_size for each agent. for example, pr2 requires max_step_size <= 0.001 for gripping task. iCub also require max_step_size <= 0.001 for walking task. otherwise you can't train a successful policy.
But pioneer works well when max_step_size=0.005. The performance for pioneer could be 5 times faster, if we set it to 0.005 rather than 0.001.

So max_step_size will differ according agents for performance reason. Currently we read max_step_size, i.e., world_time_precision, from the file agent.cfg. 

It is more natural to define step_time rather than NUM_SIMULATION_STEPS (substeps) for a task. Otherwise when agent changes, max_step_size changes, and then the time span of env step will be changed.
for example, if we set NUM_SIMULATION_STEPS (substeps)  to 100;  PR2-> pioneer; then max_step_size 0.001->0.005; control frequency 10Hz -> 2Hz 
If given steptime 0.1, then the control frequency will be fixed to 10Hz. NUM_SIMULATION_STEPS (substeps) will be caculated by steptime and max_step_size of agent.

So the model determines the max_step_size(i.e., substep_time/world_time_precision); and we should set steptime according to the task; it is indeed better to compute the NUM_SIMULATION_STEPS (substeps) from the two variables.

Reply:
Thanks for the clarifications Jiangtao.  Since step_time is the elapsed time per each step, and env class (playground) already has step_time, we should probably remove step_time from being a class member variable of the Tasks.  We can send in a world obj to task_specific_observation() function to compute speed.  The rest of the class probably doesn't need to know about step_time?

Reply:
Thanks for the kindly suggestion! step_time is removed from being a class member variable of the Tasks now. I added a get_step_time() function to environment. Please review again :-)